proposals/0000-unsatisfiable.rst

```diff
@@ -0,0 +1,599 @@
+are well-specified, and fit well with GHC's constraint-based type inference
+algorithm.  This means it should be simpler and more predictable than
+``TypeError``.  The examples above demonstrate that the issues raised in the
+Motivation have now got more principled solutions.
+
+``Unsatisfiable`` does not subsume ``TypeError`` entirely, because the former is
+restricted to kind ``Constraint``, whereas the latter is kind-polymorphic.  Thus
+there may be situations where ``TypeError`` is preferable, and both will remain
+available for use.  Correspondingly, this proposal does not lead to significant
+backwards incompatibility.
+
+
+Interaction with deferred type errors
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As usual, when ``-fdefer-type-errors`` is enabled, unsolved constraint errors
+will be deferred to runtime by the compiler automatically generating a
+dictionary that throws a runtime exception if evaluated.  The message attached
+to the exception contains the type error that was deferred (including its source
+position).
+
+This proposal does not change this behaviour; a use of ``Unsatisfiable`` merely
+changes the message that is produced.  For example::
+
+    {-# OPTIONS_GHC -fdefer-type-errors #-}
+
+    f :: Unsatisfiable (Text "Blah blah") => a -> a
+    f x = x+1
+
+    main = print (f True)
+
+Compiling this program will emit a warning corresponding to the deferred type
+error.  Executing it will throw a runtime exception like this::
+
+    *** Exception: Unsatisfiable.hs:24:15: error:
+        • Blah blah
+            arising from a use of ‘f’
+        • In the first argument of ‘print’, namely ‘(f True)’
+          In the expression: print (f True)
+          In an equation for ‘main’: main = print (f True)
+    (deferred type error)
+
+
+Costs and Drawbacks
+-------------------
+
+This is yet another feature, and will require some implementation effort, but it
+should not require extensive changes to GHC's existing constraint solving
+behaviour.
+
+The differences between ``Unsatisfiable`` and ``TypeError`` may be subtle for
+novice users, but the increased convenience of ``Unsatisfiable`` for more
+advanced users defining custom type errors in libraries seems worth it.
+
+
+Alternatives
+------------
+
+There have been various requests for more powerful alternatives to
+``TypeError``.  This proposal is deliberately simple. `Proposal #59
+<https://github.com/ghc-proposals/ghc-proposals/pull/59>`_ and `proposal #278
+<https://github.com/ghc-proposals/ghc-proposals/pull/278>`_ were more ambitious
+attempts to improve custom type errors, but both have been abandoned due to
+their complexity.
+
+`#18978 <https://gitlab.haskell.org/ghc/ghc/-/issues/18978>`_ suggests
+introducing ``Annotate :: ErrorMessage -> Constraint -> Constraint`` where
+``Annotate e c`` renders the message ``e`` if the constraint ``c`` cannot be
+solved, and is equivalent to ``c`` otherwise.  This is similar to
+``WithMessage`` from `proposal #59
```

Comment:
I think that having `Annotate e c` as a wanted is the same as having `Unsatifisiable e => c` as a wanted (it's different as a given, I think). It it exist, it should probably have the same restrictions.